core/src/test/java/apoc/refactor/GraphRefactoringTest.java

```diff
@@ -994,5 +999,170 @@ public void testRefactorCategorizeNoDups() {
+                    assertEquals("KNOWS", thirdRel.getType().name());
+                });
+    }
+
+    @Test
+    public void testMergeNodeShouldNotCreateSelfRelationshipsWithCircularPath() {
+        db.executeTransactionally("CREATE (a:TestNode {a:'a'})-[:TEST_REL]->(b:TestNode {a:'b'})-[:TEST_REL]->(c:TestNode {a:'c'})\n" +
+                "WITH a, c CREATE (c)-[:TEST_REL]->(a);");
+        testCall(db, "MATCH (n:TestNode) WITH collect(n) as nodes CALL apoc.refactor.mergeNodes(nodes, {mergeRels: true, produceSelfRel: false}) yield node return node",
+                (r) -> {
+                    Node node = (Node) r.get("node");
+                    assertFalse(node.getRelationships().iterator().hasNext());
+                });
+    }
+
+    @Test
+    public void testMergeNodeShouldCreateSelfRelationshipsWithCircularPath() {
+        db.executeTransactionally("CREATE (a:TestNode {a:'a'})-[:TEST_REL]->(b:TestNode {a:'b'})-[:TEST_REL]->(c:TestNode {a:'c'})\n" +
+                "WITH a, c CREATE (c)-[:TEST_REL]->(a);");
+        testCall(db, "MATCH (n:TestNode) WITH collect(n) as nodes CALL apoc.refactor.mergeNodes(nodes, {mergeRels: true, produceSelfRel: true}) yield node return node",
+                (r) -> {
+                    Node node = (Node) r.get("node");
+                    Iterator<Relationship> relIterator = node.getRelationships().iterator();
+                    assertSelfRel(relIterator.next());
+                    assertFalse(relIterator.hasNext());
+                });
+    }
+
+    @Test
+    public void testMergeNodeShouldNotCreateSelfRelationshipsWithPathWithOtherRels() {
+        db.executeTransactionally("CREATE (a:One)-[:TEST_REL1]->(b:Two)-[:TEST_REL2]->(c:Three)\n" +
+                "WITH b, c CREATE (b)-[:ASD]->(q:Four), (b)-[:ZXC]->(w:Five) WITH b, c CREATE (b)-[:QWE]->(c)");
+        testCall(db, "match (a:One),(b:Two),(c:Three) with [a,b,c] as nodes CALL apoc.refactor.mergeNodes(nodes, {mergeRels: true, produceSelfRel: false}) yield node return node",
+                (r) -> {
+                    Node node = (Node) r.get("node");
+                    List<String> relNodeList = IteratorUtils.toList(node.getRelationships().iterator()).stream().
+                            map(i-> i.getType().name()).collect(Collectors.toList());
+                    assertThat(relNodeList, Matchers.containsInAnyOrder("ASD", "ZXC"));
+                });
+    }
+
+    @Test
+    public void testMergeNodeShouldCreateSelfRelationshipsWithPathWithOtherRels() {
+        db.executeTransactionally("CREATE (a:One)-[:TEST_REL1]->(b:Two)-[:TEST_REL2]->(c:Three)\n" +
+                "WITH b, c CREATE (b)-[:ASD]->(q:Four), (b)-[:ZXC]->(w:Five) WITH b, c CREATE (b)-[:QWE]->(c)");
+        testCall(db, "match (a:One),(b:Two),(c:Three), ()-[rel:ASD]->() with [a,b,c] as nodes, rel CALL apoc.refactor.mergeNodes(nodes, {mergeRels: true}) yield node return node, rel",
```

Comment:
```suggestion
        testCall(db, "match (a:One),(b:Two),(c:Three) with [a,b,c] as nodes CALL apoc.refactor.mergeNodes(nodes, {mergeRels: true}) yield node return node",
```
nit: `rel` does not seem needed here